src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageFactory.cs

```diff
@@ -0,0 +1,79 @@
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
+{
+    public class DefaultPageFactory : IPageFactoryProvider
+    {
+        private readonly IPageActivatorProvider _pageActivator;
+        private readonly IModelMetadataProvider _modelMetadataProvider;
+        private readonly RazorPagePropertyActivator.PropertyValueAccessors _propertyAccessors;
+
+        public DefaultPageFactory(
+            IPageActivatorProvider pageActivator,
+            IModelMetadataProvider metadataProvider,
+            IUrlHelperFactory urlHelperFactory,
+            IJsonHelper jsonHelper,
+            DiagnosticSource diagnosticSource,
+            HtmlEncoder htmlEncoder,
+            IModelExpressionProvider modelExpressionProvider)
+        {
+            _pageActivator = pageActivator;
+            _modelMetadataProvider = metadataProvider;
+            _propertyAccessors = new RazorPagePropertyActivator.PropertyValueAccessors
+            {
+                UrlHelperAccessor = context => urlHelperFactory.GetUrlHelper(context),
+                JsonHelperAccessor = context => jsonHelper,
+                DiagnosticSourceAccessor = context => diagnosticSource,
+                HtmlEncoderAccessor = context => htmlEncoder,
+                ModelExpressionProviderAccessor = context => modelExpressionProvider,
+            };
+        }
+
+        public virtual Func<PageContext, object> CreatePageFactory(CompiledPageActionDescriptor actionDescriptor)
+        {
+            if (!typeof(Page).GetTypeInfo().IsAssignableFrom(actionDescriptor.PageTypeInfo))
+            {
+                throw new InvalidOperationException(Resources.FormatActivatedInstance_MustBeAnInstanceOf(
+                    _pageActivator.GetType().FullName,
+                    typeof(Page).FullName));
+            }
+
+            var activatorFactory = _pageActivator.CreateActivator(actionDescriptor);
+            var modelType = actionDescriptor.ModelTypeInfo?.AsType() ?? actionDescriptor.PageTypeInfo.AsType();
+            var propertyActivator = new RazorPagePropertyActivator(
+                actionDescriptor.PageTypeInfo.AsType(),
+                modelType,
+                _modelMetadataProvider,
+                _propertyAccessors);
+
+            return (context) =>
+            {
+                var page = activatorFactory(context) as Page;
+                page.PageContext = context;
+                propertyActivator.Activate(page, context);
+                return page;
+            };
+        }
+
+        public Action<PageContext, object> CreatePageDisposer(CompiledPageActionDescriptor descriptor)
+        {
+            if (descriptor == null)
+            {
+                throw new ArgumentNullException(nameof(descriptor));
+            }
+
+            return _pageActivator.CreateDisposer(descriptor);
```

Comment:
Let's make sure it's possible/allowed for these to return null if they aren't needed.